source/connection_monitor.c

```diff
@@ -0,0 +1,200 @@
+            case AWSCRT_STAT_CAT_SOCKET: {
+                struct aws_crt_statistics_socket *socket_stats = (struct aws_crt_statistics_socket *)stats_base;
+                bytes_read = socket_stats->bytes_read;
+                bytes_written = socket_stats->bytes_written;
+                break;
+            }
+
+            case AWSCRT_STAT_CAT_HTTP1: {
+                struct aws_crt_statistics_http1 *http1_stats = (struct aws_crt_statistics_http1 *)stats_base;
+                pending_read_interval_ns = http1_stats->pending_incoming_stream_ns;
+                pending_write_interval_ns = http1_stats->pending_outgoing_stream_ns;
+                break;
+            }
+
+            default:
+                break;
+        }
+    }
+
+    struct aws_channel *channel = context;
+
+    if (impl->options.minimum_throughput_bytes_per_second == 0 ||
+        impl->options.minimum_throughput_failure_threshold_in_seconds == 0) {
+        return;
+    }
+
+    /*
+     * All early-out/negative execution paths reset the failure count to zero.  Keep a copy of the current count for the
+     * remaining path.
+     */
+    uint32_t current_failure_count = impl->consecutive_throughput_failures;
+    impl->consecutive_throughput_failures = 0;
+
+    uint64_t pending_io_time_ns = 0;
+    if (aws_add_u64_checked(pending_read_interval_ns, pending_write_interval_ns, &pending_io_time_ns)) {
+        AWS_LOGF_ERROR(AWS_LS_IO_CHANNEL, "id=%p: io interval summation overflow", (void *)channel);
+        return;
+    }
+
+    if (pending_io_time_ns == 0) {
+        return;
+    }
+
+    uint64_t total_bytes = 0;
+    if (aws_add_u64_checked(bytes_read, bytes_written, &total_bytes)) {
+        AWS_LOGF_ERROR(AWS_LS_IO_CHANNEL, "id=%p: io bytes summation overflow", (void *)channel);
+        return;
+    }
+
+    uint64_t bytes_per_second = 0;
+    if (aws_mul_u64_checked(total_bytes, (uint64_t)AWS_TIMESTAMP_NANOS, &bytes_per_second)) {
+        AWS_LOGF_INFO(AWS_LS_IO_CHANNEL, "id=%p: io throughput overflow calculation", (void *)channel);
+        return;
+    }
+
+    bytes_per_second /= pending_io_time_ns;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: channel throughput - %" PRIu64 " bytes per second",
+        (void *)channel,
+        bytes_per_second);
+
+    if (bytes_per_second >= impl->options.minimum_throughput_bytes_per_second) {
+        return;
+    }
+
+    /*
+     * We failed the throughput check.  Restore and increment the failure count and then check if the failure threshold
+     * has been crossed.
+     */
+    impl->consecutive_throughput_failures = current_failure_count + 1;
+    AWS_LOGF_INFO(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: Channel low throughput warning.  Currently %u consecutive failures",
+        (void *)channel,
+        impl->consecutive_throughput_failures);
+
+    if (impl->consecutive_throughput_failures < impl->options.minimum_throughput_failure_threshold_in_seconds) {
+        return;
```

Comment:
Seems like it we'd get better behavior if we used the actual time elapsed between ticks.

Kind of like how games **try** to tick at 60FPS, but you get weird behavior if you write your game code assuming it's running at a constant tick-rate, instead of coding against actual `float delta`

instead of just counting 1,2,3 each time one of these intervals failed, you could sum the length of time they'd been failing for. So if you failed on the first tick, but you were only reading/writing a fraction of that time, then you wouldn't exceed the `minimum_throughput_failure_threshold_in_sec` because you'd only been failing for, say, 100ms so far.

Reply:
An approach like this might make the issues you were seeing with 1 second thresholds go away.

Reply:
I dunno, I'd rather force multiple samples unconditionally.  Using actual time elapsed, any single long hiccup (1 bad callback for example) will shut down the connection.  In my update I've renamed the config parameter away from a notion of time.

Reply:
Entire approach reworked